Test/presentation

yasserahmed10 opened the pull request:
This pull request introduces significant enhancements to the application, focusing on session management, audit logging, user management, and dependency injection. The most notable changes include the addition of a `SessionManager` for user role handling, the implementation of comprehensive audit logging features, and the restructuring of user-related UI classes for better modularity.

### Session Management
* Added a `SessionManager` class to manage the current user's role, enabling role-based access control throughout the application (`src/main/kotlin/di/SessionManager.kt`).
* Updated the `main` function to set the current user role to `ADMIN` during initialization (`src/main/kotlin/Main.kt`).

### Audit Logging Features
* Introduced new UI classes for managing and viewing audit logs:
  - `AddAuditLogUI` for creating new audit log entries (`src/main/kotlin/presentation/audit/AddAuditLogUI.kt`).
  - `ViewAuditLogsByProjectUI` for viewing logs by project (`src/main/kotlin/presentation/audit/ViewAuditLogsByProjectUI.kt`).
  - `ViewAuditLogsByTaskUI` for viewing logs by task (`src/main/kotlin/presentation/audit/ViewAuditLogsByTaskUI.kt`).
  - `AuditUI` as a central interface for audit log management (`src/main/kotlin/presentation/audit/AuditUI.kt`).
* Integrated `AuditUI` into the main console UI (`src/main/kotlin/presentation/PlanMateConsoleUI.kt`). [[1]](diffhunk://#diff-bb21e7bd0ad2a740cda54deb46278335bbb18b2c56d8b6e4a032d1f81c738217R3-R14) [[2]](diffhunk://#diff-bb21e7bd0ad2a740cda54deb46278335bbb18b2c56d8b6e4a032d1f81c738217L51-R57)

### User Management Enhancements
* Refactored and modularized user-related UI classes by moving them to a dedicated `presentation.user` package:
  - `CreateUserUI`, `DeleteUserUI`, `GetAllUserUI`, `GetUserByIdUI`, and `AssignProjectToUserUI` were renamed and relocated. [[1]](diffhunk://#diff-4966b72f752c21f374259a05dea9471fab6f55f0d5dc228f57533132f9cba62cL1-R8) [[2]](diffhunk://#diff-375871eaf9d5f15fa53dd90ef232ea0bc0648460ed71db785dacc1ad2d95b77dL1-R6) [[3]](diffhunk://#diff-014dc5be219dfe8f392bd3ea8802b7e6b7290055982c992c7cb140ce4cc3781dL1-R1) [[4]](diffhunk://#diff-8e5f43af9003b4566efe2ffeec0f024191247541845afb984e3f79c15d7f8c38L1-R5) [[5]](diffhunk://#diff-21fbc23587c8774678ec0ae4610144c7f146950ebc01512867ff517af577f4b6R1-R11)
* Introduced a new `UserUI` class to act as a central interface for user management (`src/main/kotlin/presentation/user/UserUI.kt`).
* Updated dependency injection in `presentationModule` to include user and audit-related components (`src/main/kotlin/di/presentationModule.kt`). [[1]](diffhunk://#diff-77e0a8aba2cc1d69e354262b86d661a594584f6558585d37771fe580b6390ac2R4-R32) [[2]](diffhunk://#diff-77e0a8aba2cc1d69e354262b86d661a594584f6558585d37771fe580b6390ac2R42-R84)

### Dependency Injection Updates
* Modified `presentationModule` to register `SessionManager` as a singleton and updated UI components to use dependency injection for session and role management (`src/main/kotlin/di/presentationModule.kt`). [[1]](diffhunk://#diff-77e0a8aba2cc1d69e354262b86d661a594584f6558585d37771fe580b6390ac2R4-R32) [[2]](diffhunk://#diff-77e0a8aba2cc1d69e354262b86d661a594584f6558585d37771fe580b6390ac2R42-R84)

### Other Changes
* Made the `users` list in `AuthenticationRepositoryImpl` public to allow external access (`src/main/kotlin/data/repositories/AuthenticationRepositoryImpl.kt`).
* Updated tests in `AuthenticationRepositoryImplTest` to align with the recent changes (`src/test/kotlin/data/repositories/AuthenticationRepositoryImplTest.kt`). [[1]](diffhunk://#diff-ad40152bba550f72422de4a506a3973059a64f4d99fcefd1f999cf583d5c23d2R7-R19) [[2]](diffhunk://#diff-ad40152bba550f72422de4a506a3973059a64f4d99fcefd1f999cf583d5c23d2L27-R29)

yasserahmed10 committed:
refactor: make users public

yasserahmed10 committed:
refactor: make users public

yasserahmed10 committed:
fix: resolve some issues

yasserahmed10 committed:
fix: resolve some issues

yasserahmed10 committed:
fix: resolve sum issues

yasserahmed10 committed:
refactor: make outputs public

yasserahmed10 committed:
refactor: make outputs public

yasserahmed10 committed:
refactor: make outputs public

yasserahmed10 committed:
test: add test cases for LoginUserUI

yasserahmed10 committed:
test: add test cases for LoginUserUI

yasserahmed10 committed:
test: add test cases for LoginUserUI

yasserahmed10 committed:
test: add test cases for RegisterAdminUI

yasserahmed10 committed:
test: add test cases for RegisterAdminUI

yasserahmed10 committed:
test: add test cases for RegisterAdminUI

yasserahmed10 committed:
test: add test cases for CreateProjectUI

yasserahmed10 committed:
test: add test cases for CreateProjectUI

yasserahmed10 committed:
test: add test cases for CreateProjectUI

yasserahmed10 committed:
test: add test cases for DeleteProjectUI

yasserahmed10 committed:
test: add test cases for DeleteProjectUI

yasserahmed10 committed:
test: add test cases for DeleteProjectUI

yasserahmed10 committed:
test: add test cases for GetAllProjectsUI

yasserahmed10 committed:
test: add test cases for GetAllProjectsUI

yasserahmed10 committed:
test: add test cases for GetAllProjectsUI

yasserahmed10 committed:
test: add test cases for ProjectsUITest

yasserahmed10 committed:
test: add test cases for ProjectsUITest

yasserahmed10 committed:
test: add test cases for ProjectsUITest

yasserahmed10 committed:
test: add test cases for UpdateProjectUI

yasserahmed10 committed:
test: add test cases for UpdateProjectUI

yasserahmed10 committed:
test: add test cases for UpdateProjectUI

yasserahmed10 committed:
fix: create a package for user

yasserahmed10 committed:
fix: create a package for user

yasserahmed10 committed:
fix: create a package for user

yasserahmed10 committed:
fix: create a package for user

yasserahmed10 committed:
Merge remote-tracking branch 'origin/test/Presentation' into test/Presentation

yasserahmed10 committed:
test: add test cases for CreateUserUI

yasserahmed10 committed:
test: add test cases for CreateUserUI

yasserahmed10 committed:
remove unuseful comment

yasserahmed10 committed:
remove unuseful comment

yasserahmed10 committed:
tests: add test cases for DeleteUserUI

yasserahmed10 committed:
tests: add test cases for DeleteUserUI

yasserahmed10 committed:
tests: add test cases for GetAllUserUI

yasserahmed10 committed:
tests: add test cases for GetAllUserUI

yasserahmed10 committed:
tests: add test cases for AssignProjectToUser

yasserahmed10 committed:
tests: add test cases for AssignProjectToUser

yasserahmed10 committed:
tests: add test cases for GetUserByIdUI

yasserahmed10 committed:
tests: add test cases for GetUserByIdUI

yasserahmed10 committed:
tests: add test cases for PlanMateConsoleUI

yasserahmed10 committed:
tests: add test cases for PlanMateConsoleUI

yasserahmed10 committed:
Merge remote-tracking branch 'origin/test/Presentation' into test/Presentation

# Conflicts:
#	src/main/kotlin/presentation/user/CreateUserUI.kt
#	src/main/kotlin/presentation/user/DeleteUserUI.kt

Commit:
Implement Audit UI

Commit:
remove unnecessary comments

baraa0abd commented on src/test/kotlin/presentation/user/GetUserByIdUITest.kt:
```diff
@@ -0,0 +1,64 @@
+        every { consoleIO.read() } returns invalidUuid
+
+        // When
+        getUserByIdUI.invoke()
```
Great job

yasserahmed10 replied:
thanks

yasserahmed10 committed:
tests: enhance PlanMateConsoleUITest with additional user and audit UI navigation tests

yasserahmed10 committed:
Merge remote-tracking branch 'origin/feature/ui' into test/Presentation

Commit:
Add session manager for get and set user role types

Commit:
Add unassigned injection parameters ViewAuditLogsByTaskUI , ViewAuditLogsByProjectUI , AuditUI , AddAuditLogUI

Commit:
Change from user role to session manager to get user role

Commit:
Add user role initialization before koin ui start

yasserahmed10 committed:
Merge remote-tracking branch 'origin/feature/ui' into test/Presentation

yasserahmed10 committed:
tests: add tests for AddAuditLogUI

yasserahmed10 committed:
tests: refactor input sequences in AddAuditLogUITest for clarity

yasserahmed10 committed:
fix: correct function calls in AuditUI and add  tests for user interactions

yasserahmed10 committed:
tests: add unit tests for ViewAuditLogsByProjectUI

yasserahmed10 committed:
tests: add unit tests for ViewAuditLogsByTaskUI

yasserahmed10 committed:
tests: add unit tests for UserUI

yasserahmed10 committed:
tests: add unit tests for AuthenticationUI

yasserahmed10 committed:
tests: remove commented-out tests from AuthenticationUITest

yasserahmed10 committed:
tests: enhance AssignProjectToUserUITest with session manager integration and error handling

yasserahmed10 committed:
tests: add tests for ConsoleIOImpl

AimanYosofiUlt commented on src/test/kotlin/presentation/auth/AuthenticationUITest.kt:
```diff
@@ -0,0 +1,131 @@
+
+
+//    @Test
+//    fun `should display menu options`() {
```
I think you shouldn't keep any commented code, please remove it

yasserahmed10 replied:
I have removed it already in  another commit after that

AimanYosofiUlt commented on src/test/kotlin/presentation/auth/AuthenticationUITest.kt:
```diff
@@ -0,0 +1,131 @@
+    }
+
+//    @Test
+//    fun `should exit when option 3 is selected`() {
```
Here too...

yasserahmed10 replied:
I have removed it in another commit after that

AimanYosofiUlt commented on src/test/kotlin/presentation/auth/AuthenticationUITest.kt:
```diff
@@ -0,0 +1,131 @@
+//    }
+
+    @Test
+    fun `should display error message for invalid input`() {
```
keep the (should-when) convention , it's good practice, like you added above in line 67 for example 👍

yasserahmed10 replied:
Done

AimanYosofiUlt reviewed (dismissed):
Nice work bro 👍

yasserahmed10 committed:
tests: improve error message descriptions in AuthenticationUITest

yasserahmed10 reviewed (commented):
I have done what is requested to change

yasserahmed10 committed:
tests: refactor AuthenticationRepositoryImplTest to use mocking and improve user registration tests

yasserahmed10 committed:
tests: refactor CsvHandlerTest to use temporary files and improve test coverage

yasserahmed10 committed:
tests: add unit tests for AuditEntity to verify constructor, equality, and copy behavior

yasserahmed10 committed:
tests: add unit tests for TasksRepositoryImpl to verify task management functionality

yasserahmed10 committed:
tests: add unit tests for TaskStateRepositoryImpl to verify task state management functionality

yasserahmed10 committed:
tests: add unit test for DeleteTaskStateUseCase to verify exception handling on deletion failure

yasserahmed10 committed:
tests: update AddTaskStateUseCase to accept titles of length 1 to 100 and add boundary condition tests

yasserahmed10 committed:
tests: refactor AddAuditLogUseCaseTest to improve structure and add new test cases for log handling

Moha-ALjabouri commented on src/main/kotlin/di/presentationModule.kt:
```diff
@@ -27,6 +39,48 @@ val presentationModule = module {
+    factoryOf(::ViewAuditLogsByProjectUI)
+    factoryOf(::ViewAuditLogsByTaskUI)
+    factoryOf(::AuditUI)
     factoryOf(::PlanMateConsoleUI)
```
great job

yasserahmed10 replied:
thanks bro

yasserahmed10 committed:
tests: refactor AddAuditLogUseCaseTest to improve structure and add new test cases for log handling

yasserahmed10 committed:
tests: refactor UserRepositoryImplTest for improved readability and structure

AimanYosofiUlt commented on src/main/kotlin/logic/usecases/task_state/AddTaskStateUseCase.kt:
```diff
@@ -20,6 +20,6 @@ class AddTaskStateUseCase(
 
     private fun isValidTitle(title: String): Boolean {
-        return title.isNotBlank() && title.length in 1..100
+        return title.isNotBlank() && title.length <= 100
```
Why should it be valid only if the length is lower than 100?

yasserahmed10 replied:
It's common practice in software design. This constraint helps:  
1-Ensure data fits within database column limits
2-Keep the UI clean with properly displaying titles
3- Prevent excessive memory usage
4-Maintain readability for users.
@AimanYosofiUlt

AimanYosofiUlt commented on src/main/kotlin/presentation/audit/ViewAuditLogsByTaskUI.kt:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    private fun readUUIDInput(prompt: String): UUID {
```
I think it should be part of the ConsoleIO, not a responsibility of this class

yasserahmed10 committed:
tests: enhance UserCsvParser to handle empty and blank project IDs with additional test cases

Hasa9nAli reviewed (dismissed):
Keep going Yasser

yasserahmed10 committed:
ui: update input handling in AddAuditLogUI to use string comparison for user choices

aabukmail commented:
@yasserahmed10 
Excellent work, Yasser. You will be appointed as the tester for our team.

yasserahmed10 committed:
tests: add unit tests for LoggedInUser class to verify user properties and equality

yasserahmed10 committed:
tests: refactor LoggedInUserTest to SessionManagerTest and enhance user session management tests

yasserahmed10 committed:
tests: add unit tests for LoggedInUser properties and project IDs

yasserahmed10 committed:
tests: add unit tests for LoggedInUser properties and project IDs

yasserahmed10 committed:
tests: add test cases for MapperKtTest.kt

yasserahmed10 committed:
tests: enhance ProjectsRepositoryImplTest with additional scenarios and verifications

baraa0abd reviewed (dismissed):
great effort , well done 🎉

Hasa9nAli committed:
refactor parameter name of menuLoop function and change the text form show projects to Project Management

Hasa9nAli committed:
edit shown message of projects and options

Hasa9nAli committed:
fix error of presentation di and rename test parameter name

Commit:
Edit login use case by changing the way of parsing data

Hasa9nAli committed:
refactor and add
1️⃣- update project name
2️⃣- add new task
3️⃣- get all tasks
4️⃣- edit task
5️⃣- get task states
6️⃣- add task state
7️⃣- delete task
8️⃣- delete task state
9️⃣- assign specific user to project
🔟- remove user form project
1️⃣1️⃣- back to previous list

Hasa9nAli committed:
Merge remote-tracking branch 'origin/test/Presentation' into test/Presentation

Commit:
modifying userui by adding getUserByIdUI , getAllUserU

Commit:
Merge remote-tracking branch 'origin/test/Presentation' into test/Presentation

yasserahmed10 committed:
refactor LoginUseCaseTest to improve user authentication tests and handle edge cases

yasserahmed10 committed:
add getUserByIdUI and update UserUI tests for user retrieval options

yasserahmed10 committed:
refactor GetAllProjectsUITest to improve verification of project output

yasserahmed10 committed:
refactor GetAllUserUITest for improved user output formatting and error handling

yasserahmed10 committed:
refactor UpdateProjectUITest to enhance task management functionality and improve user interaction

yasserahmed10 committed:
refactor PlanMateConsoleUITest to improve menu navigation and error handling

AimanYosofiUlt reviewed (dismissed):
Good Job

AimanYosofiUlt reviewed (dismissed):
Good Job
